zrml/neo-swaps/src/math.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright 2023 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{BalanceOf, Config, Error};
+use core::marker::PhantomData;
+use fixed::FixedU128;
+use hydra_dx_math::transcendental::{exp, ln};
+use sp_runtime::{DispatchError, SaturatedConversion};
+use typenum::U80;
+
+type Fractional = U80;
+type Fixed = FixedU128<Fractional>;
+
+pub(crate) trait MathOps<T: Config> {
+    fn calculate_swap_amount_out_for_buy(
+        reserve: BalanceOf<T>,
+        amount_in: BalanceOf<T>,
+        liquidity: BalanceOf<T>,
+    ) -> Result<BalanceOf<T>, DispatchError>;
+    fn calculate_swap_amount_out_for_sell(
+        reserve: BalanceOf<T>,
+        amount_in: BalanceOf<T>,
+        liquidity: BalanceOf<T>,
+    ) -> Result<BalanceOf<T>, DispatchError>;
+    fn calculate_spot_price(
+        reserve: BalanceOf<T>,
+        liquidity: BalanceOf<T>,
+    ) -> Result<BalanceOf<T>, DispatchError>;
+    fn calculate_reserves_from_spot_prices(
+        amount: BalanceOf<T>,
+        spot_prices: Vec<BalanceOf<T>>,
+    ) -> Result<(BalanceOf<T>, Vec<BalanceOf<T>>), DispatchError>;
+}
+
+pub(crate) struct Math<T>(PhantomData<T>);
+
+impl<T: Config> MathOps<T> for Math<T> {
+    fn calculate_swap_amount_out_for_buy(
+        reserve: BalanceOf<T>,
+        amount_in: BalanceOf<T>,
+        liquidity: BalanceOf<T>,
+    ) -> Result<BalanceOf<T>, DispatchError> {
+        let reserve = reserve.saturated_into();
+        let amount_in = amount_in.saturated_into();
+        let liquidity = liquidity.saturated_into();
+        detail::calculate_swap_amount_out_for_buy(reserve, amount_in, liquidity)
+            .map(|result| result.saturated_into())
```

Comment:
The return value of `detail::calculate_swap_amount_out_for_buy` is a `u128`, so the conversion won't saturate if `Balance` is `u128`. It will saturate if `Balance` has fewer bits. Don't do that, I guess...

Jokes aside, much of the math in zrml-swaps depends on `Balance` being at least `u128` (all those `saturated_into()` calls before and after using `fixed::b*` calls). From the practical perspective, this is fine as we use `Balance = u128` and changing our balance type is the last thing we want right now.

From a software engineering perspective, the problem is the intermediate step of `u128`. We can eliminate that through proper trait bounds and probably get rid of most of the `saturated_into()` calls. This will be part of the full implementation of AMM 2.0.